apps/test/unit/promiseChainTest.js

```diff
@@ -0,0 +1,225 @@
+
+    sequence.push('_');
+
+    const testCode = Promise.resolve()
+        .then(() => {
+          sequence.push('a');
+          clock.tick(1);
+        })
+        .then(() => {
+          sequence.push('b');
+        })
+        .then(() => {
+          sequence.push('c');
+          clock.tick(1);
+        })
+        .then(() => {
+          sequence.push('d');
+        });
+
+    Promise.all([promiseChainBeingTested, testCode]).then(() => {
+      expect(sequence).to.deep.equal(['_', 'a', 'A', 'b', 'c', 'B', 'd', 'C']);
+      done();
+    }).catch(done);
+  });
+
+  it(`especially when one chain has a mix of immediate and async promises.`, done => {
+    const promiseChainBeingTested = Promise.resolve()
+        .then(() => {
+          sequence.push('A');
+        })
+        .then(() => {
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('B');
+            resolve();
+          }, 1));
+        })
+        .then(() => {
+          sequence.push('C');
+        })
+        .then(() => {
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('D');
+            resolve();
+          }, 1));
+        });
+
+    sequence.push('_');
+
+    const testCode = Promise.resolve()
+        .then(() => {
+          sequence.push('a');
+        })
+        .then(() => {
+          sequence.push('b');
+          clock.tick(1);
+        })
+        .then(() => {
+          sequence.push('c');
+        })
+        .then(() => {
+          sequence.push('d');
+        })
+        .then(() => {
+          sequence.push('e');
+          clock.tick(1);
+        });
+
+    Promise.all([promiseChainBeingTested, testCode]).then(() => {
+      expect(sequence).to.deep.equal(['_', 'A', 'a', 'b', 'B', 'c', 'C', 'd', 'e', 'D']);
+      done();
+    }).catch(done);
+  });
+
+
+  it(`By wrapping _real_ setTimeout our test can yield to the promise chain.`, done => {
+    Promise.resolve()
+        .then(() => {
+          sequence.push('A');
+        })
+        .then(() => {
+          sequence.push('B');
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('C');
+            resolve();
+          }, 1));
+        })
+        .then(() => {
+          sequence.push('D');
+          return new Promise(resolve => setTimeout(() => {
+            sequence.push('E');
+            resolve();
+          }, 1));
+        })
+        .then(() => {
+          sequence.push('F');
+        });
+
+    sequence.push('_');
+
+    // This causes more nesting, but otherwise less complicated test code
+    // that's somewhat less tied to implementation
+    yieldToPromiseChain(() => {
+      sequence.push('a');
+      clock.tick(1);
+      yieldToPromiseChain(() => {
+        sequence.push('b');
+        clock.tick(1);
+        yieldToPromiseChain(() => {
+          sequence.push('c');
+          expect(sequence).to.deep.equal(['_', 'A', 'B', 'a', 'C', 'D', 'b', 'E', 'F', 'c']);
```

Comment:
would be nice for the expected order here to be more predictable, e.g. `['_', 'A', 'B', 'b', 'C', 'D', 'd', 'E', 'F', 'f']` or `['_', 'A', 'B', 'c', 'C', 'D', 'e', 'E', 'F', 'g']`